### PR TITLE
[QHC-697] Adding Calibration checkpoints

### PR DESCRIPTION
**Pull request by GuillermoAbadLopez**
## TODOs of PR:
- [x] Implement the main idea of the `checkpoints` algorithm
- [x] Implement Comparison of `fidelity` in output dictionary and values in `node.check_value`
- [ ] Test this, with concrete mock notebooks
- [x] Write quick docstrings 
- [ ] Cover functionality with tests
- [ ] Write better docstrings

Also some other checks:
- [x] Make sure is backwards compatible, if no checkpoints are passed
- [ ] Make sure weird cases indeed work correctly, like green checkpoints after red ones... 

-----
NEW TODOs:
- [ ] Do test of full flow run_automaitc_calibration, with the permutations of V and X, for both checkpoints
- [ ] End the unit test of diagnose with  the checkpoints 👌

### CODE REVIEW

File: src/qililab/calibration/calibration_controller.py [hunk 1]
```diff
@@ -235,6 +235,119 @@ def run_automatic_calibration(self) -> dict[str, dict]:
...
+        Args:
+            node (CalibrationNode): The node where we want to start the `calibration_all()` on. Normally you would want
+                this node to be the furthest node in the calibration graph.
+        """
+        logger.info("WORKFLOW: Calibrating all %s.\n", node.node_id)
+
+        # If diagnose has found a checkpoint bad, start the calibration just after the last passed checkpoint before that bad one.
+        # O - [V] - [O] - [V] - [0] - [X] - [ ] - [ ] - [ ] - ... we leave the next ones empty (.), after finding the first bad
+        # checkpoint, so that we can just find the first [V], going from right to left in ``calibrate_all()`` calls and start there.
+        if node.check_point_passed is True:
```
Comment (IsaacLA12) [inline src/qililab/calibration/calibration_controller.py:250]:
```suggestion
        if node.check_point_passed:
```

File: src/qililab/calibration/calibration_controller.py [hunk 2]
```diff
@@ -235,6 +235,119 @@ def run_automatic_calibration(self) -> dict[str, dict]:
...
+        # If you want to start the calibration from the start again, just decrease the `drift_timeout` or remove the executed files!
+        if not node.been_calibrated:
+            if node.previous_timestamp is None or self._is_timeout_expired(node.previous_timestamp, self.drift_timeout):
+                self.calibrate(node)
+                self._update_parameters(node)
+
+            node.been_calibrated = True
+        # After passing this block `node.been_calibrated` will always be True, so it will not be recalibrated again.
+
+    def diagnose(self, node: CalibrationNode) -> bool:
```
[the remark this answers is not in the document]
Comment (IsaacLA12) [reply src/qililab/calibration/calibration_controller.py:266]: Maybe `calibrate_check_points` or `analyzing_check_points` or `check_state_check_points`

[the remark this answers is not in the document]
Comment (GuillermoAbadLopez) [reply src/qililab/calibration/calibration_controller.py:266]: Or `diagnose_check_points` 🙃

**Comment by IsaacLA12** [comment]
Remember to add a test of the `diagnose` method as I believe is not covered rn

File: src/qililab/calibration/calibration_controller.py [hunk 3]
```diff
@@ -235,6 +235,119 @@ def run_automatic_calibration(self) -> dict[str, dict]:
...
+            node (CalibrationNode): The node to diagnose.
+
+        Returns:
+            bool: Wether the diagnose process has finished or not.
+        """
+        logger.info("WORKFLOW: Diagnosing  %s.\n", node.node_id)
+        diagnose_finished = False
+
+        for n in self._dependencies(node):
+            diagnose_finished = self.diagnose(n)
```
[the remark this answers is not in the document]
Comment (GuillermoAbadLopez) [reply src/qililab/calibration/calibration_controller.py]: Discussed, and changed the logic accordingly

**Comment by IsaacLA12** [comment]
After careful review, testing it with edge cases, hand to hand with @GuillermoAbadLopez, I believe it is good to go.
Leaving my aprove but adding software testing is needed to pass codecov and also test it on HW when available.

**Review by IsaacLA12 — APPROVED** [review]

**Comment by GuillermoAbadLopez** [comment]
Not needed anymore
